lib/internal/modules/esm/resolve.js

```diff
@@ -506,9 +408,49 @@ function packageExportsResolve(
   throwExportsNotFound(packageSubpath, packageJSONUrl, base);
 }
 
+function readIfFile(path) {
+  let fd;
+  try {
+    fd = openSync(path, 'r');
+  } catch {
+    return undefined;
+  }
+  try {
+    if (!fstatSync(fd).isFile()) return undefined;
+    return readFileSync(fd, 'utf8');
+  } finally {
+    closeSync(fd);
+  }
+}
+
+function catchInvalidPackage(error, path, base) {
+  if (StringPrototypeIncludes(error.message, 'Error parsing')) {
+    throw new ERR_INVALID_PACKAGE_CONFIG(
+      join(path, 'package.json'), fileURLToPath(base), false);
```

Comment:
We should carry through the parse message from `error.message` as well here for context (`error.message.slice(('Error parsing ' + path + ': ').length)` sort of thing maybe)